### DISCUSSION TITLE
How to read a billion-scale edge list

### DISCUSSION THREAD

**Discussion by han-jiang-etsy** [networkit/networkit#1291] [Q&A]
Hi team,

I am experimenting a super big graph with ~ 14 B edge list. The size of the txt file is about 400G. A few questions
1. How to efficiently read the txt file, and is there a better file format in this case?
2. Is it possible if I partition the file into multiple files and networkit reads each and combines them in memory (not sure if networkit supports add graph to an existing one). Will this be faster than reading a single file
Thanks in advance

**Comment by hmeyerhenke** [comment]
0) Make sure you have enough RAM. Sounds like you need around 300-500 GB, give or take.
1) NetworKit's binary format would certainly be more efficient in terms of disk space and I/O time. But you would probably have to read the file once for its conversion anyway.
2) Although I am not 100% sure, I doubt that we have such a mechanism in place. Plus I am not sure if there is an advantage. You could think of implementing something similar in Python, though. When reading additional graph files, each edge you see is added to the already existing graph. But this will cost performance. And ultimately you need enough RAM anyway.

**Comment by han-jiang-etsy** [comment]
@hmeyerhenke thanks for the reply. Can you elaborate more on the binary format and what api to do that. I am using apache Beam [write_to_text function](https://beam.apache.org/releases/pydoc/2.2.0/apache_beam.io.textio.html#apache_beam.io.textio.WriteToText). so I assume I need some way to convert the txt file on disk to binary file, or I need to apply networkit api in one of the Beam step.

**Comment by hmeyerhenke** [comment]
Take a look at the io module, BinaryReader and BinaryWriter. @angriman and @avdgrinten would be able to tell you more about the format.

**Comment by han-jiang-etsy** [comment]
yeah, I am looking for something like beam.Map(json.dumps) to encode the text into binary
below is the sample code  
```
with beam.Pipeline(options=pipeline_options) as pipeline:
        with tft_beam.Context(temp_dir=custom_gcs_temp_location):
            raw_data = pipeline | "QueryTable" >> beam.io.ReadFromBigQuery(
                query=query, use_standard_sql=True
            )
            raw_data = raw_data | "ParseValue" >> beam.Map(parse_values)
            # raw_data | "print" >> beam.Map(print)
            _ = raw_data \
                | "FormatJson" >> beam.Map(json.dumps) \
                | "WriteTextToJson" >> beam.io.textio.WriteToText(
                os.path.join(output_root, "part"), num_shards=num_shards
            )
```
